Add MeshModifiers syntax

kyleeswanson opened the pull request:
refs Issue #27671

## Reason
To add folders just for MeshModifiers - UOs that modify MooseMesh

## Impact
Will help to keep moose/framework more organized as capabilities grow 


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

Commit:
Created MeshModifiers folder in framework - Issue #27671

Organizing MeshModifiers folders - Issue #27671

Removed changes.diff - Issue #27671

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2369331/) on 46922c1 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28314/site)

This comment will be updated on new commits.

GiudGiud commented on changes.diff:
should not be included

GiudGiud commented on framework/doc/content/source/meshmodifiers/TimedSubdomainModifier.md:
redo this change but with git mv ?
here it shows as deleted then re-added

kyleeswanson replied:
I see it as renamed in my commit history at the terminal

roystgnr replied:
Looks like this is a Github issue?  https://github.com/orgs/community/discussions/43177

Commit:
Added syntax to Moose.C - Issue #27671

Commit:
Added AddMeshModifiersAction.C and AddMeshModifiersAction.h

Changed _problem->addMeshModifiers to _problem->addUserObject in AddMeshModifiersAction.C

Commit:
Updated registerMooseObjectTask add_mesh_modfier from MeshModifier to UserObject - Issue #27671

Commit:
Updated test.i cli_arg to include MeshModifiers

Commit:
Added correct dependencies to Moose.C

Commit:
Updating .mds to work with MeshModifiers and build website correctly - Issue #27671

GiudGiud commented on framework/doc/content/syntax/MeshModifiers/index.md:
```diff
@@ -0,0 +1,18 @@
+
+The `MeshModifiers` system is designed to be able to modify the connectivity of the MooseMesh.
+
+## Postprocessing with MeshModifiers
+
+For now, the mesh divisions can only be output using a [MeshDivisionAux.md].
+We are planning to be able to compute averages, integrals, and various reductions
+on mesh divisions, please stay tuned.
+
+## Transferring with MeshModifiers
+
+This feature is a work in progress.
+
+!alert note
+Please let us know on [GitHub Discussions](https://github.com/idaholab/moose/discussions)
+how you are using the `MeshModifiers` system so we can include your techniques on this page!
```
```suggestion
```
not relevant for mesh modifiers

see the syntax page for user objects please. there are some commands we want to add to list all the objects

GiudGiud commented on framework/include/actions/AddMeshModifiersAction.h:
```diff
@@ -0,0 +1,22 @@
+
+#include "MooseObjectAction.h"
+
+class AddMeshModifiersAction : public MooseObjectAction
```
see other actions, add a docstring here
it s a little C++ comment

GiudGiud commented on framework/src/actions/AddMeshModifiersAction.C:
```diff
@@ -0,0 +1,32 @@
+AddMeshModifiersAction::validParams()
+{
+  InputParameters params = MooseObjectAction::validParams();
+  params.addClassDescription("Add a MeshModifiers object to the simulation.");
```
```suggestion
  params.addClassDescription("Add a MeshModifier object to the simulation.");
```

GiudGiud commented on framework/src/actions/AddMeshModifiersAction.C:
```diff
@@ -0,0 +1,32 @@
+void
+AddMeshModifiersAction::act()
+{
+  _problem->addUserObject(_type, _name, _moose_object_pars);
```
```suggestion
  // MeshModifiers are user objects in the backend
  _problem->addUserObject(_type, _name, _moose_object_pars);
```

GiudGiud commented on framework/src/base/Moose.C:
```diff
@@ -188,6 +188,8 @@ addActionTypes(Syntax & syntax)
   registerMooseObjectTask("add_control",                  Control,                   false);
   registerMooseObjectTask("add_partitioner",              MoosePartitioner,          false);
 
+  registerMooseObjectTask("add_mesh_modifier",            UserObject,             false);
```
pls move this line next to the userObject one

GiudGiud commented on framework/src/base/Moose.C:
```diff
@@ -233,6 +235,8 @@ addActionTypes(Syntax & syntax)
   registerTask("check_integrity_early_physics", false);
   registerTask("setup_quadrature", true);
 
+  registerTask("mesh_modifiers", false);
```
same

kyleeswanson replied:
I don't see a registerTask for userObjects

GiudGiud commented on framework/src/base/Moose.C:
```diff
@@ -611,6 +615,10 @@ associateSyntaxInner(Syntax & syntax, ActionFactory & /*action_factory*/)
 
+  // MeshModifiers
+  registerSyntax("AddMeshModifiersAction", "MeshModifiers/*");
+  syntax.registerSyntaxType("MeshModifiers/*", "MeshModifiersName");
```
let's move these next to the UO syntax

GiudGiud reviewed (changes requested):
looking good this made great progress in very little time. well done!

Commit:
Making changes Guillaume requested - Issue #27671

Commit:
updated filepath in .md - Issue #2767

Commit:
Adding ActivatesElementUserObjectBase, ActivateElementsCoupled, and ActivateElementsByPath to MeshModifiers

Updating docu for MeshModifiers

Commit:
Created AddMeshModifiers.md - Issue #27671

commented out syntax parametes in AddMeshModifiers.md

GiudGiud committed:
Create a new base for MeshModifier, to avoid user objects being created in the new syntax
Add main syntax page link
Add listings to MeshMod syntax page

Remove bad docstring in header
Improve doco wording

Co-authored-by: Gary (Tianchen) Hu <thu@anl.gov>

Update framework/src/meshmodifiers/TimedSubdomainModifier.C

Co-authored-by: Gary (Tianchen) Hu <thu@anl.gov>

hugary1995 commented on framework/doc/content/syntax/MeshModifiers/index.md:
```diff
@@ -0,0 +1,14 @@
+# MeshModifiers
+
+The `MeshModifiers` system is designed to be able to modify the [MooseMesh.md] during the
```
```suggestion
The `MeshModifiers` system is designed to modify the [MooseMesh.md] during the
```

hugary1995 commented on framework/doc/content/source/meshmodifiers/ActivateElementsUserObjectBase.md:
```diff
@@ -2,4 +2,4 @@
 The `ActivateElementsUserObjectBase` class is for activating (adding) an element by moving the element from an "inactive" subdomain to the "active" subdomain. There are two metrics for activating an element, i.e., by +function path+ and by +coupled variable value+, which are implemented in the classes [ActivateElementsByPath](/ActivateElementsByPath.md) and [ActivateElementsCoupled](/ActivateElem ...
 
-This user object updates the boundary information due to the addition of elements across subdomains. User can save this boundary information by passing the changed boundary name to  `expand_boundary_name` in the input block.  Note that current implementation does not de-activate an element once the element is activated.
+This MeshModifier updates the boundary information due to the addition of elements across subdomains. User can save this boundary information by passing the changed boundary name to  `expand_boundary_name` in the input block.  Note that current implementation does not de-activate an element once the element is activated.
```
```suggestion
This mesh modifier updates the boundary information due to the addition of elements across subdomains. User can save this boundary information by passing the changed boundary name to  `expand_boundary_name` in the input block.  Note that current implementation does not de-activate an element once the element is activated.
```

hugary1995 commented on framework/include/actions/AddMeshModifiersAction.h:
```diff
@@ -0,0 +1,31 @@
+class AddMeshModifiersAction : public MooseObjectAction
+{
+public:
+  /**
+   * Class constructor
+   * @param params Parameters for this Action
+   */
```
```suggestion
```

hugary1995 commented on framework/src/meshmodifiers/TimedSubdomainModifier.C:
```diff
@@ -48,7 +48,7 @@ TimedSubdomainModifier::validParams()
   params.addClassDescription(
       "Modify element subdomain ID of entire subdomains for given points "
-      "in time. This userobject only runs on the undisplaced mesh, and it will "
+      "in time. This meshmodifier only runs on the undisplaced mesh, and it will "
```
```suggestion
      "in time. This mesh modifier only runs on the undisplaced mesh, and it will "
```

hugary1995 reviewed (changes requested):
LGTM

GiudGiud commented:
review merged in the last commit

btw @hugary1995 we do persistent approvals (remains after a new commit is added) now. So you can approve when requesting minor things and let the CCB contributor merge after addressing

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2399203/) on 7a6eb42 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>bc8f02</th>
      <th colspan=4>#28314 7a6eb4</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.01%</td>
      <td>85.01%</td>
      <td>+0.00%</td>
      <td>63.01%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>105060</td>
      <td>105077</td>
      <td>+17</td>
      <td>615</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18523</td>
      <td>18525</td>
      <td>+2</td>
      <td>361</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28314/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28314/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28314/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 63.01% is less than the suggested 90.0%



This comment will be updated on new commits.
